kimchi/src/circuits/expr.rs

```diff
@@ -1396,10 +1397,19 @@ impl<F: FftField> Expr<F> {
                 evals: env.vanishes_on_last_4_rows,
             },
             Expr::Constant(x) => EvalResult::Constant(*x),
-            Expr::UnnormalizedLagrangeBasis(i) => EvalResult::Evals {
-                domain: d,
-                evals: unnormalized_lagrange_evals(env.l0_1, *i, d, env),
-            },
+            Expr::UnnormalizedLagrangeBasis(i) => {
+                let i = {
+                    if *i >= 0 {
+                        *i as u32
+                    } else {
+                        (d1_size as u32) * (d as u32) - ((-*i) as u32)
```

Comment:
TODO: the `d` factor is already accounted for in the function call below; it should be deleted here.